File: src/main/java/org/radarcns/stream/SensorStreamWorker.java
/*
 * Copyright 2017 King's College London and The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarcns.stream;

import static org.apache.kafka.streams.KeyValue.pair;
import static org.radarcns.util.StreamUtil.first;
import static org.radarcns.util.StreamUtil.second;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ThreadLocalRandom;
import java.util.function.Function;
import java.util.stream.Collectors;
import javax.annotation.Nonnull;
import org.apache.avro.Schema;
import org.apache.avro.specific.SpecificRecord;
import org.apache.kafka.streams.KafkaStreams;
import org.apache.kafka.streams.KeyValue;
import org.apache.kafka.streams.StreamsBuilder;
import org.apache.kafka.streams.StreamsConfig;
import org.apache.kafka.streams.kstream.KStream;
import org.apache.kafka.streams.kstream.TimeWindows;
import org.radarbase.stream.collector.AggregateListCollector;
import org.radarbase.stream.collector.NumericAggregateCollector;
import org.radarcns.kafka.AggregateKey;
import org.radarcns.kafka.ObservationKey;
import org.radarcns.stream.aggregator.AggregateList;
import org.radarcns.stream.aggregator.NumericAggregate;
import org.radarcns.util.Monitor;
import org.radarcns.util.RadarSingleton;
import org.radarcns.util.RadarUtilities;
import org.radarcns.util.serde.RadarSerdes;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Abstraction of a Kafka Stream.
 * @param <K> input key type.
 * @param <V> input value type.
 */
public abstract class SensorStreamWorker<K extends SpecificRecord, V extends SpecificRecord>
        extends AbstractStreamWorker {
    @SuppressWarnings("PMD.LoggerIsNotStaticFinal")
    private final Logger monitorLog;
    private Collection<ScheduledFuture<?>> monitors;

    protected final RadarUtilities utilities = RadarSingleton.getInstance().getRadarUtilities();

    public SensorStreamWorker() {
        this.streams = null;
        this.monitors = null;
        this.monitorLog = LoggerFactory.getLogger(getClass());
    }

    /**
     * Create a Kafka Stream builder. This implementation will create a stream from given
     * input topic to given output topic. It monitors the amount of messages that are read.
     */
    protected KeyValue<ScheduledFuture<?>, KafkaStreams> createBuilder(StreamDefinition def) {

        StreamsBuilder builder = new StreamsBuilder();

        KStream<?, ?> stream = implementStream(def, builder.stream(def.getInputTopic().getName()));

        ScheduledFuture<?> future = null;
        if (monitorLog != null) {
            Monitor monitor = new Monitor(monitorLog, "records have been read from "
                    + def.getInputTopic() + " to " + def.getOutputTopic());
            stream = stream.peek((k, v) -> monitor.increment());
            future = master.addMonitor(monitor);
        }

        if (def.getOutputTopic() != null) {
            stream.to(def.getOutputTopic().getName());
        }

        return pair(future, new KafkaStreams(builder.build(), getStreamProperties(def)));
    }

    /**
     * @return Properties for a Kafka Stream
     */
    protected Properties getStreamProperties(@Nonnull StreamDefinition definition) {
        StringBuilder clientIdBuilder = new StringBuilder(100);
        clientIdBuilder.append(getClass().getName())
                .append('-')
                .append(allConfig.getBuildVersion());
        TimeWindows window = definition.getTimeWindows();
        if (window != null) {
            clientIdBuilder.append('-')
                    .append(window.sizeMs)
                    .append('-')
                    .append(window.advanceMs);
        }

        Properties props = kafkaProperty.getStreamProperties(clientIdBuilder.toString(), config,
                DeviceTimestampExtractor.class);
        long interval = (long)(ThreadLocalRandom.current().nextDouble(0.75, 1.25)
                * definition.getCommitInterval().toMillis());
        props.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG,
                String.valueOf(interval));

        return props;
    }

    protected Map<String, ?> getStreamPropertiesMap(@Nonnull StreamDefinition definition) {
        Map<String, Object> map = new HashMap<>();
        Properties properties = getStreamProperties(definition);
        properties.forEach((k, v) -> map.put(k.toString(), v));
        return map;
    }

    /**
     * Defines the stream computation.
     */
    protected abstract KStream<?, ?> implementStream(StreamDefinition definition,
            @Nonnull KStream<K, V> kstream);

    /**
     * Starts the stream and notify the StreamMaster.
     */
    @Override
    public List<KafkaStreams> createStreams() {
        List<KeyValue<ScheduledFuture<?>, KafkaStreams>> streamBuilders = getStreamDefinitions()
                .map(this::createBuilder)
                .collect(Collectors.toList());

        monitors = streamBuilders.stream()
                .map(first())
                .collect(Collectors.toList());

        return streamBuilders.stream()
                .map(second())
                .collect(Collectors.toList());
    }

    @Override
    protected void doCleanup() {
        if (monitors != null) {
            monitors.forEach(f -> f.cancel(false));
            monitors = null;
        }
    }

    protected final KStream<AggregateKey, NumericAggregate> aggregateNumeric(
            @Nonnull StreamDefinition definition, @Nonnull KStream<ObservationKey, V> kstream,
            @Nonnull String fieldName, @Nonnull Schema schema) {
        return kstream.groupByKey()
                .windowedBy(definition.getTimeWindows())
                .aggregate(
                        () -> new NumericAggregateCollector(fieldName, schema,
                                config.isUseReservoirSampling()),
                        (k, v, valueCollector) -> valueCollector.add(v),
                        RadarSerdes.materialized(definition.getStateStoreName(),
                            RadarSerdes.getInstance(allConfig.getSchemaRegistryPaths())
                                    .getNumericAggregateCollector(
                                            getStreamPropertiesMap(definition), false)))
                .toStream()
                .map(utilities::numericCollectorToAvro);
    }

    protected final KStream<AggregateKey, NumericAggregate> aggregateCustomNumeric(
            @Nonnull StreamDefinition definition, @Nonnull KStream<ObservationKey, V> kstream,
            @Nonnull Function<V, Double> calculation, @Nonnull String fieldName) {
        return kstream.groupByKey()
                .windowedBy(definition.getTimeWindows())
                .aggregate(
                        () -> new NumericAggregateCollector(fieldName,
                                config.isUseReservoirSampling()),
                        (k, v, valueCollector) -> valueCollector.add(calculation.apply(v)),
                        RadarSerdes.materialized(definition.getStateStoreName(),
                            RadarSerdes.getInstance(allConfig.getSchemaRegistryPaths())
                                    .getNumericAggregateCollector(
                                            getStreamPropertiesMap(definition), false)))
                .toStream()
                .map(utilities::numericCollectorToAvro);
    }

    protected final KStream<AggregateKey, AggregateList> aggregateFields(
            @Nonnull StreamDefinition definition, @Nonnull KStream<ObservationKey, V> kstream,
            @Nonnull String[] fieldNames, @Nonnull Schema schema) {

        return kstream.groupByKey()
                .windowedBy(definition.getTimeWindows())
                .aggregate(
                        () -> new AggregateListCollector(fieldNames, schema,
                                config.isUseReservoirSampling()),
                        (k, v, valueCollector) -> valueCollector.add(v),
                        RadarSerdes.materialized(definition.getStateStoreName(),
                            RadarSerdes.getInstance(allConfig.getSchemaRegistryPaths())
                                    .getAggregateListCollector(
                                            getStreamPropertiesMap(definition), false)))
                .toStream()
                .map(utilities::listCollectorToAvro);
    }

    @Override
    public String toString() {
        return getClass().getSimpleName();
    }
}

File: src/main/kotlin/org/radarbase/kotlin/stream/SensorStreamWorker.kt
package org.radarbase.kotlin.stream

import org.apache.avro.Schema
import org.apache.avro.specific.SpecificRecord
import org.apache.kafka.streams.KafkaStreams
import org.apache.kafka.streams.KeyValue
import org.apache.kafka.streams.StreamsBuilder
import org.apache.kafka.streams.StreamsConfig
import org.apache.kafka.streams.kstream.KStream
import org.radarbase.kotlin.util.Monitor
import org.radarbase.kotlin.util.RadarSingletonFactory
import org.radarbase.kotlin.util.RadarUtilities
import org.radarbase.kotlin.util.StreamUtil
import org.radarbase.kotlin.util.serde.RadarSerdes
import org.radarbase.stream.collector.AggregateListCollector
import org.radarbase.stream.collector.NumericAggregateCollector
import org.radarcns.kafka.AggregateKey
import org.radarcns.kafka.ObservationKey
import org.radarcns.stream.aggregator.AggregateList
import org.radarcns.stream.aggregator.NumericAggregate
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.ThreadLocalRandom
import java.util.stream.Collectors

abstract class SensorStreamWorker<K : SpecificRecord, V : SpecificRecord> : AbstractStreamWorker() {
    private val monitorLog = LoggerFactory.getLogger(javaClass)
    private var monitors: MutableCollection<ScheduledFuture<*>>? = null

    protected val utilities: RadarUtilities = RadarSingletonFactory.radarUtilities

    init {
        this.streams = null
        this.monitors = mutableListOf<ScheduledFuture<*>>()
    }

    internal fun createBuilder(def: StreamDefinition): KeyValue<ScheduledFuture<*>?, KafkaStreams> {
        val monitor: Monitor?
        val future: ScheduledFuture<*>?
        if (monitorLog != null) {
            monitor = Monitor(monitorLog, "records have been read from ${def.inputTopic} to ${def.outputTopic}")
            future = master.addMonitor(monitor as org.radarbase.util.Monitor)
        } else {
            monitor = null
            future = null
        }

        val builder = StreamsBuilder()

        val kstream = implementStream(
            def,
            builder.stream<K, V>(def.inputTopic.name)
                .map { k, v ->
                    monitor?.increment()
                    KeyValue.pair(k, v)
                }
        )

        val outputTopicName = def.outputTopic.name
        kstream.to(outputTopicName)

        val properties = getStreamProperties(def)
        return KeyValue.pair(future, KafkaStreams(builder.build(), properties))
    }

    internal fun getStreamProperties(definition: StreamDefinition): Properties {
        val localClientId = buildString {
            append(javaClass.name)
            append("-")
            append(allConfig.buildVersion)
            definition.timeWindows?.let {
                append("-")
                append(it.sizeMs)
                append("-")
                append(it.advanceMs)
            }
        }

        val props = kafkaProperty.getStreamProperties(
            localClientId,
            config,
            DeviceTimestampExtractor::class.java
        )

        val interval = (ThreadLocalRandom.current().nextDouble(0.75, 1.25) *
                definition.commitInterval.toMillis()).toLong()

        props[StreamsConfig.COMMIT_INTERVAL_MS_CONFIG] = interval.toString()

        return props
    }

    internal abstract fun implementStream(definition: StreamDefinition, kstream: KStream<K, V>): KStream<*, *>

    override fun createStreams(): List<KafkaStreams>? {
        val streamBuilders = getStreamDefinitions()
            .map { createBuilder(it) }
            .collect(Collectors.toList())

        monitors = streamBuilders.stream()
            .map(StreamUtil.first())
            .filter { it != null }
            .map { it!! }
            .collect(Collectors.toList())

        return streamBuilders.stream()
            .map(StreamUtil.second())
            .collect(Collectors.toList())
    }

    override fun doCleanup() {
        monitors?.forEach { it.cancel(false) }
        monitors = null
    }

    protected fun aggregateNumeric(
        definition: StreamDefinition,
        kstream: KStream<ObservationKey, V>,
        fieldName: String,
        schema: Schema
    ): KStream<AggregateKey, NumericAggregate> {
        return kstream.groupByKey()
            .windowedBy(definition.timeWindows)
            .aggregate(
                { NumericAggregateCollector(fieldName, schema) },
                { _, v, valueCollector -> valueCollector.add(v) },
                RadarSerdes.materialized(
                    definition.stateStoreName,
                    RadarSerdes.getInstance().getNumericAggregateCollector() as org.apache.kafka.common.serialization.Serde<NumericAggregateCollector>
                )
            )
            .toStream()
            .map(utilities::numericCollectorToAvro)
    }

    protected fun aggregateCustomNumeric(
        definition: StreamDefinition,
        kstream: KStream<ObservationKey, V>,
        calculation: (V) -> Double,
        fieldName: String
    ): KStream<AggregateKey, NumericAggregate> {
        return kstream.groupByKey()
            .windowedBy(definition.timeWindows)
            .aggregate(
                { NumericAggregateCollector(fieldName) },
                { _, v, valueCollector -> valueCollector.add(calculation(v)) },
                RadarSerdes.materialized(
                    definition.stateStoreName,
                    RadarSerdes.getInstance().getNumericAggregateCollector() as org.apache.kafka.common.serialization.Serde<NumericAggregateCollector>
                )
            )
            .toStream()
            .map(utilities::numericCollectorToAvro)
    }

    protected fun aggregateFields(
        definition: StreamDefinition,
        kstream: KStream<ObservationKey, V>,
        fieldNames: Array<String>,
        schema: Schema
    ): KStream<AggregateKey, AggregateList> {
        return kstream.groupByKey()
            .windowedBy(definition.timeWindows)
            .aggregate(
                { AggregateListCollector(fieldNames, schema, false) },
                { _, v, valueCollector -> valueCollector.add(v) },
                RadarSerdes.materialized(
                    definition.stateStoreName,
                    RadarSerdes.getInstance().getAggregateListCollector() as org.apache.kafka.common.serialization.Serde<AggregateListCollector>
                )
            )
            .toStream()
            .map(utilities::listCollectorToAvro)
    }

    override fun toString(): String {
        return javaClass.simpleName
    }
}
